Ticketing Tutorial (#16)
keypom/keypom-docs: docs/Tutorials/Advanced/ticketing/user-code.md

:::

Recall, from the [introduction](introduction.md) that your code had the following outline.

```bash
/my-ticket-app
├── contract
│   └── ...
├── frontend
│   └── components
│   │    └── scanner.js
│   │    └── qrcode.js
│   └── state
│   │    └── App.js
│   │    └── keyInfo.js
│   └── utils
│   │    └── configurations.js
│   │    └── createTickDrop.js
│   └── node_modules
│   │    └── keypom-js
│   │    └── qrcode.react
│   │    └── react-zxing
│   │    └── react
│   │    └── react-dom
│   │    └── react-router-dom
│   │    └── ...
│   └── package.json
│   └── package-lock.json
├── ...
├── package.json
├── package-lock.json
```

This tutorial will be covering the code in `app.js`, `KeyInfo.js`, and `qrcode.js`.

---

## `App.js`
### Setting Up
The primary purpose of `app.js` is to display the different states of the claim page. This will involve getting the current key uses, and then rendering based on the value returned. 

The first step is to initialize a connection to NEAR and setup all the state variables that will be needed to render the page. 

> ```jsx reference showLineNumbers

Reviewer: This code snippet is cutoff
<img width="649" alt="image" src="https://user-images.githubusercontent.com/57506486/224809788-53b4d5ca-5eeb-4196-b994-237ef110fd54.png">

https://github.com/keypom/keypom-js/blob/0035852580c8ce848571d89f31ae47a3794414d6/docs-advanced-tutorials/ticket-app/frontend/state/App.js#L15-L59
```

> When the page is loaded, the function `setup` is called and the URL is parsed for `contractId` and `privateKey`. These will be stored in their own respective state  variables (lines 40-45) for further use. Note that the URL is split by `/` and the index of `contractId` and `privateKey` are known. When adapting this code for your own app, you may need to change those index values accordingly.

Reviewer: FYI we will need to use `url#contractId/secretKey` instead of `url/contractId/secretKey` otherwise the contract and secret key are logged on databases and search engines like google analytics.

Once the URL has been parsed and the resulting values stored, `connectNear` is called. This function handles the NEAR connection, as well as the Keypom connection and the linkdrop URL.

:::note 
When connecting to NEAR, a `BrowserLocalStorageKeyStore` is used rather than an `UnencryptedFileSystemKeyStore`. This will create a key store in the user's browser for future use. For more on key stores, see [here](https://docs.near.org/tools/near-api-js/quick-reference#key-store)
:::

With the NEAR connection established, your browser can now talk to the NEAR blockchain. The next step, is to call `initKeypom`. This initializes the SDK to allow for interactions with the Keypom smart contracts. Without it, none of the other SDK functions would work as expected. More info on the `initKeypom` function can be found [here](../../../keypom-sdk/modules#initkeypom).

> After the Keypom initialization is complete, the SDK function [`formatLinkdropUrl`](../../../keypom-sdk/modules.md#formatlinkdropurl) can be used to create the linkdrop link to embed in the QR code. To use this function, it simply needs a base URL, as well as the `privateKey` to be used. The returned linkdrop link is set to the state variable `link` and will be used to render the QR code.

Reviewer: why don't we just take the contract ID and private key and turn that into the QR code? Why use format linkdrop URL?

### Rendering

The process of rendering is simple: retrieve the key's current use number, and render the page differently based on the stages outlined in the [App Design](react-outline.md). 

> During this process, the `QrCode` and `KeyInfo` components are called, passing in the values and state variables defined in the [set up](user-code.md#setting-up). For the time being, you can think of the `KeyInfo` component as just a black box, where a `privKey`, `curUse` and other state variables are passed in, and those state variables get modified to change what is rendered. `QRCode` uses the `qrcode.react` library to display a QR code based on a string input.

Reviewer: ```suggestion
The information is passed into both a QR code component and a KeyInfo component. For the time being, you can think of the `KeyInfo` component as just a black box, where a `privKey`, `curUse` and other state variables are passed in, and those state variables get modified to change what is rendered. `QRCode` uses the `qrcode.react` library to display a QR code based on a string input.
```

The following table outlines what the page should render based on a set of conditions including the current key use. 

|    **Condition and Description**                                                                                                                                                                                                      | **Render**                                                                                                                                                                              |
|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **`curUse` is 1**: The claim page should show a QR code representing the linkdrop link. The `QrCode` component is rendered by passing in the `link` state variable created earlier.                                                   | <p align="center"> <img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-1.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>       |

Reviewer: ```suggestion
| **`curUse` is 1**: The claim page should show a QR code representing the ticket data (contract and secret key). The `QrCode` component is rendered by passing in the `link` state variable created earlier.                                                   | <p align="center"> <img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-1.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>       |
```

| **`curUse` is 2**: The first claim was done by the bouncer, and thus should give the user the option to claim their POAP. This is done with the same linkdrop link.                                                                   | <p align="center"><img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-2.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>        |
| **`curUse` is 0**: The second claim was used, and thus the key was depleted and deleted. Here, the user will be given resources to continue learning about NEAR. You can choose to show anything here.                                | <p align="center"><img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-3.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>        |

> | **`curUse` is 0 **and** splitRes[3] is '' or undefined**: `splitRes[3]` is 0 or undefined indicates the private key does not exist in the URL. This is to act as a landing page for your event, you can choose to show anything here. | <p align="center"><img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-0.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>        |

Reviewer: ```suggestion
| This should act as a landing page for your event and will be shown if any URL is invalid. | <p align="center"><img src={require("/static/img/docs/advanced-tutorials/ticketing/claim-0.png").default} width="80%" height="80%" alt="ticketing" class="rounded-corners"/></p>        |
```

The code to render can be found in the exandable section below. 

<details>
<summary>app.js rendering code</summary>
<p>

> ```jsx reference showLineNumbers

Reviewer: Let's use a switch statement here

https://github.com/keypom/keypom-js/blob/0035852580c8ce848571d89f31ae47a3794414d6/docs-advanced-tutorials/ticket-app/frontend/state/App.js#L61-L131
```

</p>
</details>

> :::tip 

Reviewer: remove

Lines 2, 23, 40, and 55 hold conditions on the current key uses, `curUse`. These conditions control what is shown to the user.
:::

### Full Code
The full code, including the set up and the rendering, can be seen below.
<details>
<summary>Full app.js code</summary>
<p>

``` jsx reference
https://github.com/keypom/keypom-js/blob/0035852580c8ce848571d89f31ae47a3794414d6/docs-advanced-tutorials/ticket-app/frontend/state/App.js#L1-L136
```

</p>
</details>

--- 

## `Qrcode.js`

`Qrcode.js` is a simple component that returns a QR code based on the string passed in. On the claim page, this is the linkdrop URL for the doorman to scan. 

The full code can be seen below. 

```js reference
https://github.com/keypom/keypom-js/blob/0035852580c8ce848571d89f31ae47a3794414d6/docs-advanced-tutorials/ticket-app/frontend/components/qrcode.js#L1-L22
```

--- 

## `Keyinfo.js`

`Keyinfo.js` is responsible for updating Keypom related state variables, `pubKey`, and `curUse`. 

> When `curUse` is 1 and the QR code is being rendered, its imparative to show the public key and key use, so the user has this information on hand in case the QR code becomes unreadable. 

Reviewer: ```suggestion
When `curUse` is 1 and the QR code is being rendered, the public key and use number is shown for educational purposes.
```
